src/response/usage_report.rs

```diff
@@ -0,0 +1,293 @@
+use std::prelude::v1::*;
+
+use core::fmt;
+
+use chrono::DateTime;
+use serde::{
+    de::{self, MapAccess, Visitor},
+    Deserialize, Deserializer,
+};
+
+mod systemtime {
+    use chrono::DateTime;
+
+    #[repr(transparent)]
+    #[derive(Debug, Clone, PartialEq, Eq)]
+    pub struct PeriodTime(pub i64);
+
+    impl<Tz: chrono::TimeZone> From<DateTime<Tz>> for PeriodTime {
+        fn from(dt: DateTime<Tz>) -> PeriodTime {
+            PeriodTime(dt.timestamp())
+        }
+    }
+}
+
+pub use systemtime::PeriodTime;
+
+#[non_exhaustive]
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum UsageReportError {
+    LimitsExceeded(u64),
+    Overflow,
+}
+
+#[derive(Debug, Clone, Deserialize, PartialEq, Eq)]
+#[serde(rename = "usage_report")]
+pub struct UsageReport {
+    pub metric: String,
+    pub period: Period,
+    pub period_start: PeriodTime,
+    pub period_end: PeriodTime,
+    pub max_value: u64,
+    pub current_value: u64,
+}
+
+impl UsageReport {
+    pub fn metric(&self) -> &str {
+        self.metric.as_ref()
+    }
+
+    pub fn period(&self) -> &Period {
+        &self.period
+    }
+
+    pub fn period_times(&self) -> (&PeriodTime, &PeriodTime) {
+        (&self.period_start, &self.period_end)
+    }
+
+    pub fn max_value(&self) -> u64 {
+        self.max_value
+    }
+
+    pub fn current_value(&self) -> u64 {
+        self.current_value
+    }
+
+    pub fn remaining(&self) -> u64 {
+        self.max_value.saturating_sub(self.current_value)
+    }
+
+    pub fn is_limited(&self) -> bool {
+        self.current_value >= self.max_value
```

Comment:
Shouldn't this be `>` ? :thinking:

Reply:
`is_limited` is defined (for me, anyway) as "can't hit again".